File: src/store/task.ts
import { reactive } from "vue";

export interface TaskProgress {
  status: "Idle" | "Running" | "Cancelled" | "Finished";
  progress: number;
  total: number;
}

type FrontendStatus = TaskProgress["status"] | "Countdown";

export const taskState = reactive<{
  status: FrontendStatus;
  progress: number;
  total: number;
}>({
  status: "Idle",
  progress: 0,
  total: 0,
});

export function setTaskState(newState: TaskProgress) {
  taskState.status = newState.status;
  taskState.progress = newState.progress;
  taskState.total = newState.total;
  if (newState.status === "Finished" || newState.status === "Cancelled") {
    setTimeout(() => {
      taskState.status = "Idle";
      taskState.progress = 0;
      taskState.total = 0;
    }, 2000);
  }
}

File: src-tauri/src/domain/models/task.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Serialize, Deserialize, TS, PartialEq, Default)]
#[ts(export)]
pub enum TaskStatus {
    #[default]
    Idle,
    Running,
    Cancelled,
    Finished,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS, Default)]
#[ts(export)]
pub struct TaskProgress {
    pub status: TaskStatus,
    pub progress: u32,
    pub total: u32,
}
